src/main/java/org/apache/commons/collections4/iterators/ExtendedIterator.java

```diff
@@ -0,0 +1,197 @@
+ * @param <T> The type of object returned from the iterator.
+ */
+public final class ExtendedIterator<T> implements Iterator<T> {
+    /**
+     * Set to <code>true</code> if this wrapping doesn't permit the use of
+     * {@link #remove()}, otherwise removal is delegated to the base iterator.
+     */
+    private final boolean removeDenied;
+
+    /**
+     * Creates an ExtendedIterator wrapped round <code>it</code>,
+     * which does not permit <code>.remove()</code>
+     * even if <code>it</code> does.
+     * @param it The Iterator to wrap.
+     * @return an Extended iterator on {@code it}
+     * @throws UnsupportedOperationException if remove() is called on the resulting iterator.
+     */
+    public static <T> ExtendedIterator<T> createNoRemove(final Iterator<T> it) {
+        return new ExtendedIterator<>(it, true);
+    }
+
+    /**
+     * Creates an ExtendedIterator wrapped round a {@link Stream}.
+     * The extended iterator does not permit <code>.remove()</code>.
+     * <p>
+     * The stream should not be used directly. The effect of doing so is
+     * undefined.
+     * </p>
+     * @param stream the Stream to create an iterator from.
+     * @return an Extended iterator on the {@code stream} iterator.
+     */
+    public static <T> ExtendedIterator<T> create(final Stream<T> stream) {
+        return new ExtendedIterator<T>(stream.iterator(), true);
+    }
+
+    /**
+     * Flattens an iterator of iterators into an Iterator over the next level values.
+     * Similar to list splicing in lisp.
+     * @param iterators An iterator of iterators.
+     * @return An iterator over the logical concatenation of the inner iterators.
+     */
+    public static <T> ExtendedIterator<T> flatten(final Iterator<Iterator<T>> iterators) {
+        return create(new LazyIteratorChain<T>() {
+
+            @Override
+            protected Iterator<? extends T> nextIterator(final int count) {
+                return iterators.hasNext() ? iterators.next() : null;
+            }
+
+        });
+    }
+
+    /**
+     * Creates an empty Extended iterator.
+     * @return An empty Extended iterator.
+     */
+    public static ExtendedIterator<?> emptyIterator() {
+        return new ExtendedIterator<>(Collections.emptyIterator(), false);
+    }
+
+    /**
+     * Create an ExtendedIterator returning the elements of <code>it</code>.
+     * If <code>it</code> is itself an ExtendedIterator, return that;
+     * otherwise wrap <code>it</code>.
+     * @param it The iterator to wrap.
+     * @return An Extended iterator wrapping {@code it}
+     */
+    public static <T> ExtendedIterator<T> create(final Iterator<T> it) {
+        return it instanceof ExtendedIterator<?>
+                ? (ExtendedIterator<T>) it
+                : new ExtendedIterator<>(it, false);
+    }
+
+    /** the base iterator that we wrap */
+    private final Iterator<? extends T> base;
```

Comment:
Extend `FilterIterator` and you don't need this and most other delegating methods.

Reply:
`setFunction()` ? This class does not track a function that I can see. 

There is just too much duplication with `FilterIterator` to avoid subclassing IMO. Having 2 almost identical classes feels confusing IMO.  

I could even see initially creating a subclass and then considering pulling up the functionality into FilterIterator. The cuurent FilterIterator already does more than a plain "Filter" in the traditional `Filter[Input|Output]Stream` model (The same model we follow Commons IO where we provide subclasses to `Filter[Input|Output]Stream` (most) extending `Proxy[Input|Output]Stream`.)

As an aside, ideally, the FilterIterator class should have been immutable, or that's what I would do if I were to write one today, but that's beside the point IMO. I say this as I do not think we want to add more mutable state.

Reply:
Hi @Claudenw 
You wrote:
> If we do this then why not expose setFunction() and something else for andThen().

Which is why I asked about `setFunction()`. Maybe I'm the one missing something? 🤔

Reply:
Hi @Claudenw,

> EnhancedIterator is an Iterator that adds Streaming like functionality

This is causing me to pause and think that adding streaming-like functionality here is the wrong way to go. Instead of adding a 2 or 3 methods now and who knows how many later, and considering that the `Stream` interface and functionality is large and rich, so instead of inventing a new vocabulary here, why not just use a `Stream`? We could instead add:

- org.apache.commons.collections4.IteratorUtils.stream(Iterable<E>)
- org.apache.commons.collections4.IteratorUtils.stream(Iterator<E>)

This provides a bridge to the full Stream functionality, which everyone is familiar with, no need to add and learn new custom methods like "addTo", "map", and so on.


Details: Something like
```java
    /**
     * Creates a stream on the given Iterable.
     *
     * @param <E> the type of elements in the Iterable.
     * @param iterable the Iterable to stream or null.
     * @return a new Stream or {@link Stream#empty()} if the Iterable is null.
     * @since 4.5.0-M3
     */
    public static <E> Stream<E> stream(final Iterable<E> iterable) {
        return iterable == null ? Stream.empty() : StreamSupport.stream(iterable.spliterator(), false);
    }

    /**
     * Creates a stream on the given Iterator.
     *
     * @param <E> the type of elements in the Iterator.
     * @param iterator the Iterator to stream or null.
     * @return a new Stream or {@link Stream#empty()} if the Iterator is null.
     * @since 4.5.0-M3
     */
    public static <E> Stream<E> stream(final Iterator<E> iterator) {
        return iterator == null ? Stream.empty() : StreamSupport.stream(Spliterators.spliteratorUnknownSize(iterator, Spliterator.ORDERED), false);
    }
```